src/connection.rs

```diff
@@ -648,10 +692,67 @@ impl Connection {
 
     /// Updates [`ConnectionQualityScore`] of this [`Connection`].
     pub fn update_quality_score(&self, score: ConnectionQualityScore) {
-        if self.0.quality_score.replace(Some(score)) != Some(score) {
-            // TODO: Replace with derive?
-            #[expect(clippy::as_conversions, reason = "needs refactoring")]
-            self.0.on_quality_score_update.call1(score as u8);
+        if self.0.quality_score.replace(Some(score)) == Some(score) {
+            return;
+        }
+
+        let client_score =
+            self.new_client_quality_score(self.0.peer_state.get(), Some(score));
+
+        if let Some(client_score) = client_score {
+            self.on_quality_score_update(client_score);
         }
     }
+
+    /// Updates [`PeerConnectionState`] of this [`Connection`].
+    pub fn update_peer_state(&self, state: PeerConnectionState) {
+        if self.0.peer_state.replace(Some(state)) == Some(state) {
+            return;
+        }
+
+        let client_score = self
+            .new_client_quality_score(Some(state), self.0.quality_score.get());
+
+        if let Some(client_score) = client_score {
+            self.on_quality_score_update(client_score);
+        }
+    }
+
+    /// Creates new [`ClientConnectionQualityScore`] if it's changed.
+    fn new_client_quality_score(
+        &self,
+        state: Option<PeerConnectionState>,
+        quality_score: Option<ConnectionQualityScore>,
+    ) -> Option<ClientConnectionQualityScore> {
+        let score = match (state, quality_score) {
+            (
+                Some(
+                    PeerConnectionState::Disconnected
+                    | PeerConnectionState::Failed,
+                ),
+                _,
+            ) => Some(ClientConnectionQualityScore::Disconnected),
+            (
+                Some(
+                    PeerConnectionState::New | PeerConnectionState::Connecting,
+                ),
+                _,
+            ) => self.0.client_quality_score.get(),
+            (_, Some(quality_score)) => Some(quality_score.into()),
+            _ => None,
+        };
+
+        if self.0.client_quality_score.replace(score) == score {
+            return None;
+        }
+
+        score
+    }
+
+    /// Notifies subscriber about [`ClientConnectionQualityScore`] change.
+    fn on_quality_score_update(&self, score: ClientConnectionQualityScore) {
+        // TODO: Replace with derive?
+        #[expect(clippy::as_conversions, reason = "needs refactoring")]
+        self.0.on_quality_score_update.call1(score as u8);
+    }
```

Comment:
It can be moved inside `update_client_quality_score` I think. Main point is that we should have one place where we calculate `ClientConnectionQualityScore` and call `on_quality_score_update` callback. So calculation logic and callback calling logic of this score is not spread out across the code base.